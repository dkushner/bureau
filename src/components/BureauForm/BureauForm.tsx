import { Component, Vue, Prop, Provide } from 'vue-property-decorator'
import { VNode, CreateElement } from 'vue'
import { namespace } from 'vuex-class'
import { FormProxy } from './index'

const Bureau = namespace('bureau')

@Component
export default class BureauForm extends Vue {
  @Prop({ type: String, required: true }) name!: string
  @Prop([Array, Function]) validate!: []
  @Prop(Function) onSubmit!: () => void
  @Prop(Boolean) eagerDestroy: boolean = false
  @Prop(Boolean) keepDirty: boolean = false
  @Prop(Object) initialValues!: { [key: string]: any }

  @Bureau.Getter getRegisteredForms!: any
  @Bureau.Getter getForm!: any
  @Bureau.Getter getFormFields!: any
  @Bureau.Getter getFormField!: any
  @Bureau.Getter getFormValues!: any
  @Bureau.Mutation change!: any
  @Bureau.Mutation initialize!: any
  @Bureau.Mutation registerField!: any
  @Bureau.Mutation unregisterField!: any
  @Bureau.Mutation focus!: any
  @Bureau.Mutation blur!: any

  @Provide()
  form: FormProxy = {
    registerField: (fieldName, validators) => {
      this.registerField({ formName: this.name, fieldName })

      if (validators) {
        this.validators[fieldName] = validators
      }
    },
    unregisterField: (fieldName: string) => {
      delete this.validators[fieldName]
      this.unregisterField({ 
        formName: this.name, 
        fieldName, 
        destroy: this.eagerDestroy 
      })
    },
    change: (fieldName: string, value: any) => {
      this.change({ formName: this.name, fieldName, value })
    },
    focus: (fieldName: string) => {
      this.focus({ formName: this.name, fieldName })
    },
    blur: (fieldName: string) => {
      this.blur({ formName: this.name, fieldName })
    },
    getFieldState: (fieldName: string) => {
      return this.getFormField(this.name, fieldName)
    },
    getValue: (fieldName: string) => {
      return this.getFormValues(this.name)[fieldName]
    }
  }

  validators: { [key: string]: any } = { }

  handleSubmit () {
    this.onSubmit()
  }

  created () {
    this.initialize({
      formName: this.name,
      values: this.initialValues,
      keepDirty: this.keepDirty
    })
  }

  render (h: CreateElement): VNode {
    const bindings = {
      on: {
        submit: this.handleSubmit
      }
    }

    return (
      <form { ...bindings }>
        { this.$slots.default }
      </form>
    )
  }
 }