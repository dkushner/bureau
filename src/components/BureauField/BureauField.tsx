import { Component, Vue, Prop, Inject, Watch } from "vue-property-decorator";
import { FormProxy } from '../BureauForm'
import { CreateElement, VNode } from "vue";

@Component
export default class BureauField extends Vue {
  @Prop({ type: String, required: true }) name!: string
  @Prop([Array, Function]) validate!: []
  @Prop(Function) normalize!: any
  @Prop(Function) format!: any

  @Inject('form') form!: FormProxy

  get inputListeners() {
    const instance = this

    return Object.assign({ }, this.$listeners, {
      input (event) {
        instance.$emit('input', event)
      },
      change (event) {
        let value = event.target.value

        if (instance.normalize) {
          value = instance.normalize(instance.name, value)
        }

        instance.form.change(instance.name, value)
        instance.$emit('change', event)
      },
      focus (event) {
        instance.form.focus(instance.name)
        instance.$emit('focus', event)
      },
      blur (event) {
        instance.form.blur(instance.name)
        instance.$emit('blur', event)
      }
    })
  }

  get active() {
    const field = this.form.getFieldState(this.name)
    return field ? field.active : false
  }

  get value() {
    return this.form.getValue(this.name)
  }

  @Watch('name')
  onNameChanged (newValue: string, oldValue: string) {
    this.form.unregisterField(this.name)
    this.form.registerField(this.name, this.validate)
  }

  mounted () {
    this.form.registerField(this.name, this.validate)
  }

  beforeDestroy () {
    this.form.unregisterField(this.name)
  }

  render (h: CreateElement): VNode {
    const bindings = {
      props: {
        field: {
          value: this.value
        },
        listeners: this.inputListeners,
        form: this.form
      }
    }
    
    return (
      <slot { ...bindings }></slot>
    )
  }
}