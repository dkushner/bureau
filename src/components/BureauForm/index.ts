import BureauForm from './BureauForm'
import { FieldState } from '../../state';

export interface FormProxy {
  registerField (fieldName: string, validators: any[]): void
  unregisterField (fieldName: string): void
  change (fieldName: string, value: any): void
  focus (fieldName: string): void
  blur (fieldName: string): void
  getFieldState (fieldName: string): FieldState
  getValue (fieldName: string): any
}

export { BureauForm }
export default BureauForm