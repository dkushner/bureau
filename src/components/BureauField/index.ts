import BureauField from './BureauField'

export { BureauField }
export default BureauField 