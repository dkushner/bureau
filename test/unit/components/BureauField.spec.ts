import Vuex from 'vuex'
import { mount, createLocalVue, Wrapper } from '@vue/test-utils';
import { mockStore, Overrides } from '../../index'
import BureauField from '../../../src/components/BureauField'
import { BureauForm, FormProxy } from '../../../src/components/BureauForm';

const mockFormProxy = (overrides: Partial<FormProxy> = { }): FormProxy => ({
  blur: jest.fn(),
  change: jest.fn(),
  focus: jest.fn(),
  getFieldState: jest.fn(),
  registerField: jest.fn(),
  getValue: jest.fn(),
  unregisterField: jest.fn(),
  ...(overrides || { })
})

describe('BureauField', () => {

  test('registers the field to the form context', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const propsData = {
      name: 'test'
    }

    const form: FormProxy = mockFormProxy()

    const wrapper = mount(BureauField, {
      localVue,
      propsData,
      provide: { form },
      scopedSlots: {
        default: `
        <span slot-scope='props'>
          {{ props }}
        </span>
        `
      }
    })

    const span = wrapper.find('span')
    expect(span.exists()).toBeTruthy()
    expect(span.text()).toMatch(/form/)
    expect(span.text()).toMatch(/field/)
    expect(span.text()).toMatch(/listeners/)
    expect(form.registerField).toBeCalled()
  })

  test('emits when bound input changes', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const propsData = {
      name: 'test'
    }

    const form: FormProxy = mockFormProxy({
      getValue: jest.fn(name => name)
    })

    const wrapper = mount(BureauField, {
      localVue,
      propsData,
      provide: { form },
      scopedSlots: {
        default: `
        <span slot-scope="{ field, listeners }">
          <input type="text" :value="field.value" v-on="listeners" />
        </span>
        `
      }
    })

    wrapper.find('input').trigger('change')

    expect(form.change).toBeCalled()
    expect(form.change).toBeCalledWith('test', 'test')
    expect(wrapper.emitted()).toHaveProperty('change')
    expect(wrapper.emitted().change).toHaveLength(1)
  })

  test('emits when bound input is focused', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const propsData = {
      name: 'test'
    }

    const form: FormProxy = mockFormProxy()

    const wrapper = mount(BureauField, {
      localVue,
      propsData,
      provide: { form },
      scopedSlots: {
        default: `
        <span slot-scope="{ field, listeners }">
          <input type="text" :value="field.value" v-on="listeners" />
        </span>
        `
      }
    })

    wrapper.find('input').trigger('focus')

    expect(form.focus).toBeCalled()
    expect(form.focus).toBeCalledWith('test')
    expect(wrapper.emitted()).toHaveProperty('focus')
    expect(wrapper.emitted().focus).toHaveLength(1)
  })

  test('emits when bound input is modified', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const propsData = {
      name: 'test'
    }

    const form: FormProxy = mockFormProxy()

    const wrapper = mount(BureauField, {
      localVue,
      propsData,
      provide: { form },
      scopedSlots: {
        default: `
        <span slot-scope="{ field, listeners }">
          <input type="text" :value="field.value" v-on="listeners" />
        </span>
        `
      }
    })

    wrapper.find('input').trigger('input')
    expect(wrapper.emitted()).toHaveProperty('input')
    expect(wrapper.emitted().input).toHaveLength(1)
  })


  test('emits when bound input is blurred', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const propsData = {
      name: 'test'
    }

    const form: FormProxy = mockFormProxy()

    const wrapper = mount(BureauField, {
      localVue,
      propsData,
      provide: { form },
      scopedSlots: {
        default: `
        <span slot-scope="{ field, listeners }">
          <input type="text" :value="field.value" v-on="listeners" />
        </span>
        `
      }
    })

    wrapper.find('input').trigger('blur')

    expect(form.blur).toBeCalled()
    expect(form.blur).toBeCalledWith('test')
    expect(wrapper.emitted()).toHaveProperty('blur')
    expect(wrapper.emitted().blur).toHaveLength(1)
  })

  test('renders as expected', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const propsData = {
      name: 'test'
    }

    const form: FormProxy = mockFormProxy({
      getValue: jest.fn(name => name)
    })

    const wrapper = mount(BureauField, {
      localVue,
      propsData,
      provide: { form },
      scopedSlots: {
        default: `
        <span slot-scope="{ field, listeners }">
          <input type="text" :value="field.value" v-on="listeners" />
        </span>
        `
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})