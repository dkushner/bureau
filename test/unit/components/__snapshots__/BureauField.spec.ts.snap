// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BureauField renders as expected 1`] = `

<span>
  <input type="text">
</span>

`;
